pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -308,6 +309,155 @@ func (am *Alertmanager) SyncAndApplyConfigFromDatabase() error {
 	return nil
 }
 
+func (am *Alertmanager) TestReceivers(ctx context.Context, c apimodels.TestReceiversConfig) (*apimodels.TestReceiversResult, error) {
+	// now represents the start time of the test
+	now := time.Now()
+	testAlert := &types.Alert{
+		Alert: model.Alert{
+			Labels: model.LabelSet{
+				model.LabelName("instance"): "foo",
+			},
+			Annotations: model.LabelSet{},
+			StartsAt:    now,
+		},
+		UpdatedAt: now,
+	}
+
+	// we must set a group key as this is required by some notification channels (i.e. webhook)
+	ctx = notify.WithGroupKey(ctx, testAlert.Labels.String())
+
+	tmpl, err := am.getTemplate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get template: %w", err)
+	}
+
+	integrationsMap, err := am.buildIntegrationsMap(c.Receivers, tmpl)
+	if err != nil {
+		// The request contains invalid config for one or more contact points
+		return nil, fmt.Errorf("failed to create receivers: %w", err)
+	}
+
+	// job contains all metadata required to test a receiver
+	type job struct {
```

Comment:
Big fan of the approach here but I'd love it if could simplify it a bit more like:

1. Create the jobs
2. Return if we don't have any
3. Create channel `ch := make(chan job, len(jobs))`
4. Send all the jobs to the change and then close it:
```
	for _, j := range jobs {
		ch <- j
	}
	close(ch)
```
5. Evaluate needed concurrency (per my comment below)
6. Create the error group using the context and process the results `g, gCtx := errgroup.WithContext(ctx)` and `g.Go(func() error { for key := range ch { ... } })`

The only extra thing we'll need here is a lock through the results. I find this a bit easier than keeping track of the other n+2 goroutines but can def see this being a subjective thing and would make the expensive part of the process (notify + result aggregation) context aware.